### PR TITLE
test: Reduce end-to-end test time

### PR DESCRIPTION
**Pull request by sunchao**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Currently Java tests take over 2 hours to finish, which is very long. This increases the end-to-end time for pull requests to be processed, and decreases the developer efficiency.

Screenshot 2024-02-27 at 8 40 11 AM

## What changes are included in this PR?

This PR makes a few changes in shuffle & aggregation tests:

- The shuffle tests are now split into `CometColumnarShuffleSuite` and `CometNativeShuffleSuite`, for columnar and native shuffle respectively. This is more like a refactoring rather than changing the tests themselves
- Reduces the test time for "Comet shuffle: different data type", and only run it with native shuffle enabled. We have a separate test for columnar shuffle.
- Removes `NonFastMerge` related tests. In Spark, the fast merge feature has been true by default since 1.4, so it is not so useful to run all tests with the feature being turned off. If needed, we can add a dedicated test for it.
- Reduce the number of rows for "fix: Too many task completion listener of ArrowReaderIterator causes OOM" by 10X. This test takes extremely long time to finish.
- Reduce the time for "all types, with nulls" in `CometAggregateSuite`. 

The total time is now reduced to ~40min

### CODE REVIEW

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 1]
```diff
@@ -837,18 +759,13 @@ abstract class CometShuffleSuiteBase extends CometTestBase with AdaptiveSparkPla
 
   test("fix: Native Unsafe decimal accessors return incorrect results") {
     Seq(10, 201).foreach { numPartitions =>
-      withSQLConf(
-        CometConf.COMET_EXEC_ENABLED.key -> "true",
-        CometConf.COMET_COLUMNAR_SHUFFLE_ENABLED.key -> "true",
-        CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true") {
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:843]: Hmm, these configs are used to produce a specific situation. If they are changed, you won't fail the test but maybe it just tests nothing.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 2]
```diff
@@ -27,24 +27,24 @@ import org.apache.spark.{Partitioner, SparkConf}
...
-
   protected val fastMergeEnabled: Boolean = true
-
   protected val numElementsForceSpillThreshold: Int = 10
 
   override protected def sparkConf: SparkConf = {
     val conf = super.sparkConf
     conf
       .set(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key, adaptiveExecutionEnabled.toString)
+      .set(CometConf.COMET_EXEC_ENABLED.key, "false")
```
Comment (sunchao) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:45]: moved the common configs for columnar shuffle to the `sparkConf`.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 3]
```diff
@@ -925,193 +831,64 @@ abstract class CometShuffleSuiteBase extends CometTestBase with AdaptiveSparkPla
...
-          val shuffled3 = df.repartition(10, $"_2", $"_1")
-
-          checkCometExchange(shuffled3, 1, execEnabled)
-          checkSparkAnswer(shuffled3)
-        }
-      }
-    }
-  }
-
-  test("Comet shuffle: different data type") {
```
Comment (sunchao) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:958]: We have this which tests both columnar and native shuffle, and "Comet columnar shuffle shuffle: different data type" below which only test columnar shuffle. This PR moves the columnar shuffle from this test and moves it into `CometNativeShuffleSuite`, to reduce the test time.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 1 above)
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:843]: If you move the common configs and it should be good. There are many tests so we should be careful if no one is missed or with changed configs.

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 4]
```diff
@@ -90,13 +90,13 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     Range(0, len).map(_ => chars.charAt(r.nextInt(chars.length))).mkString
   }
 
-  private def fuzzCastFromString(chars: String, maxLen: Int, toType: DataType) {
```
Comment (sunchao) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:93]: This change is added by accident (from `make format`). I'll remove it later.

**Comment by sunchao** [comment]
cc @viirya @advancedxy

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 5]
```diff
@@ -55,7 +50,10 @@ abstract class CometShuffleSuiteBase extends CometTestBase with AdaptiveSparkPla
     super.test(testName, testTags: _*) {
       withSQLConf(
         CometConf.COMET_COLUMNAR_SHUFFLE_ASYNC_ENABLED.key -> asyncShuffleEnable.toString,
-        CometConf.COMET_EXEC_SHUFFLE_SPILL_THRESHOLD.key -> numElementsForceSpillThreshold.toString) {
+        CometConf.COMET_EXEC_SHUFFLE_SPILL_THRESHOLD.key -> numElementsForceSpillThreshold.toString,
+        CometConf.COMET_EXEC_ENABLED.key -> "false",
```
[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: I thought about this too but not exactly sure. I think it doesn't look necessary. @viirya do you remember why?

File: spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            .filter($"_1" === 1.toString)
+            .repartition(numPartitions, $"_1", $"_2")
+            .sortWithinPartitions($"_1")
+          checkSparkAnswerAndOperator(df)
+        }
+      }
+    }
+  }
+
+  test("fix: comet native shuffle with binary data") {
```
[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala:188]: Sure. It seems we don't have binary type in the Parquet table, but only FixedLengthByteArray. We can add it.

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala [hunk 7]
```diff
@@ -537,23 +537,22 @@ class CometAggregateSuite extends CometTestBase with AdaptiveSparkPlanHelper {
               withSQLConf(CometConf.COMET_BATCH_SIZE.key -> batchSize.toString) {
 
                 // Test all combinations of different aggregation & group-by types
-                (1 to 4).foreach { col =>
-                  (1 to 14).foreach { gCol =>
+                (1 to 14).foreach { gCol =>
+                  (1 to 4).foreach { col =>
```
[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala]: Right, this only covers integer types. The other types like float/double, decimal are covered by other tests in the same suite. We could add them here but it might cause explosion in the total test time.

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala [hunk 8]
```diff
@@ -537,23 +537,22 @@ class CometAggregateSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
                     withView("v") {
                       sql(s"CREATE TEMP VIEW v AS SELECT _g$gCol, _$col FROM tbl ORDER BY _$col")
                       checkSparkAnswer(s"SELECT _g$gCol, FIRST(_$col) FROM v GROUP BY _g$gCol")
```
[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala]: Good idea. Let me try this.

**Comment by sunchao** [comment]
Thanks. Updated the PR.

> TPC-DS run now is the bottleneck, I think we can reduce it in a follow up PR.

Yes it's easy to break it up.

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala [hunk 9]
```diff
@@ -537,23 +537,18 @@ class CometAggregateSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
-                      s"SELECT _g$gCol, MIN(_$col), MAX(_$col) FROM tbl GROUP BY _g$gCol")
-                    checkSparkAnswer(s"SELECT _g$gCol, AVG(_$col) FROM tbl GROUP BY _g$gCol")
+                (1 to 14).foreach { gCol =>
+                  withView("v") {
+                    sql(s"CREATE TEMP VIEW v AS SELECT _g$gCol, _1, _2, _3, _4 " +
+                      "FROM tbl ORDER BY _1, _2, _3, _4")
+                    checkSparkAnswer(s"SELECT _g$gCol, FIRST(_1), FIRST(_2), FIRST(_3), " +
+                      s"FIRST(_4), LAST(_1), LAST(_2), LAST(_3), LAST(_4) FROM v GROUP BY _g$gCol")
                   }
+                  checkSparkAnswer(s"SELECT _g$gCol, SUM(_1), SUM(_2), COUNT(_3), COUNT(_4), " +
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala:547]: Are columns `_1`, `_2`... different to each other? If so, seems this removes `SUM(_3)` and `SUM(_4)` and also misses some `SUM(DISTINCT  xx)`? Because previously it tests all columns with all aggregate expressions.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: Hmm? Did we have this before? Looks like it is added by this PR, no?

[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: Oh, there are many tests for columnar shuffle disabling it.

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala (hunk 9 above)
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala:547]: Yes, the code path is pretty much the same though, so I thought it is fine to lose some precision here.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: I think some are intentional as some bugs only happen if comet exec is disabled. Some may not necessary, and possibly because it simplifies the cases (no native operators).

[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: Yea, I forgot why we did that.

[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:54]: I see, I'll keep it then.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 3 above)
Comment (viirya) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:977]: I think by purpose these tests focus on different aspects. For example, "different data type" tests are for shuffling on different data types especially.

They might be some overlapping when we look them now. At beginning, we want to have much more test coverage as possible to reduce missing corner cases.

Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:977]: Oops you are right. I got confused when looking at the `execEnabled`.  It's interesting that for this test case,  the check `checkCometExchange(shuffled, 1, true)` can pass regardless of whether `execEnabled` is true or false. Looks like native shuffle can be enabled even if native execution is not.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 10]
```diff
@@ -913,407 +822,162 @@ abstract class CometShuffleSuiteBase extends CometTestBase with AdaptiveSparkPla
...
-        assert(metrics.contains("shuffleWriteTime"))
-        assert(metrics("shuffleWriteTime").value > 0)
-      }
+      assert(metrics.contains("shuffleWriteTime"))
+      assert(metrics("shuffleWriteTime").value > 0)
     }
   }
 }
 
-class CometAsyncShuffleSuite extends CometShuffleSuiteBase {
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:1264]: Yea, async is only meaningful for columnar shuffle. We don't need run native shuffle test cases with async.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala [hunk 11]
```diff
@@ -913,407 +822,162 @@ abstract class CometShuffleSuiteBase extends CometTestBase with AdaptiveSparkPla
...
-
-  // TODO: this test takes ~5mins to run, we should reduce the test time.
-  // Because this test takes too long, we only have it in `CometShuffleSuite`.
-  test("fix: Too many task completion listener of ArrowReaderIterator causes OOM") {
-    withSQLConf(
-      CometConf.COMET_EXEC_ENABLED.key -> "true",
-      CometConf.COMET_BATCH_SIZE.key -> "1",
-      CometConf.COMET_COLUMNAR_SHUFFLE_ENABLED.key -> "false",
-      CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true") {
-      withParquetTable((0 until 1000000).map(i => (1, (i + 1).toLong)), "tbl") {
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:1301]: I remember I set a big number because the bug only happens for many rows. We probably can reduce the number just no sure which is a proper one.

File: spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+      checkSparkAnswer(shuffled2)
+
+      val shuffled3 = df.repartition(10, $"_2", $"_1")
+
+      checkCometExchange(shuffled3, 1, true)
+      checkSparkAnswer(shuffled3)
+    }
+  }
+
+  test("columnar shuffle: single partition") {
```
Comment (viirya) [inline spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala:104]: Do we need to change test name? It is native shuffle suite.

File: spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala (hunk 3 above)
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:977]: Oh I see. I'll update and keep the `execEnabled`

File: spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala (hunk 12 above)
Comment (sunchao) [reply spark/src/test/scala/org/apache/comet/exec/CometNativeShuffleSuite.scala:104]: Yea let me update this.

**Comment by sunchao** [comment]
Merged, thanks!
